app/src/main/java/com/example/livedata/LoginFragment.kt

```diff
@@ -0,0 +1,58 @@
+package com.example.livedata
+
+import android.arch.lifecycle.Observer
+import android.arch.lifecycle.ViewModelProviders
+import android.databinding.DataBindingUtil
+import android.os.Bundle
+import android.support.v4.app.Fragment
+import android.view.LayoutInflater
+import android.view.View
+import android.view.ViewGroup
+import android.widget.EditText
+import com.example.livedata.databinding.FragmentLoginBinding
+
+
+class LoginFragment : Fragment() {
+    private lateinit var loginViewModel: LoginViewModel
+    private lateinit var binding: FragmentLoginBinding
+
+    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
+        super.onCreate(savedInstanceState)
+        loginViewModel = ViewModelProviders.of(this).get(LoginViewModel::class.java)
+        binding = DataBindingUtil.inflate(inflater, R.layout.fragment_login, container, false)
+        binding.let {
+            it.lifecycleOwner = this
+            it.loginViewModel = loginViewModel
+            return it.root
+        }
+    }
+
+    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
+        super.onViewCreated(view, savedInstanceState)
+        bindViewModel()
+    }
+
+    private fun bindViewModel() {
+        loginViewModel.user.observe(this, Observer { user ->
+            binding.run {
+                user?.let {
+                    when {
+                        it.emailAddress.isEmpty() -> setErrorOnField(emailAddress, R.string.empty_email_error)
```

Comment:
Shouldn't all this logic be in view model? You can expose an error live data field that emits what kind of error to show and the fragment only knows about displaying error with the info vm emits.